Applications/CausalLM/models/gemma3/gemma3_causallm.cpp

```diff
@@ -0,0 +1,278 @@
+     withKey("packed", "false")}));
+
+  auto att_layer =
+    createAttention(layer_id, INIT_SEQ_LEN, NUM_HEADS, HEAD_DIM,
+                    "layer" + std::to_string(layer_id) + "_attention_norm",
+                    "layer" + std::to_string(layer_id) + "_attention_norm",
+                    "layer" + std::to_string(layer_id) + "_attention_norm");
+  layers.insert(layers.end(), att_layer.begin(), att_layer.end());
+
+  layers.push_back(createLayer(
+    "rms_norm", {withKey("name", "layer" + std::to_string(layer_id) +
+                                   "_post_attention_norm"),
+                 withKey("input_layers",
+                         "layer" + std::to_string(layer_id) + "_attention_out"),
+                 withKey("epsilon", std::to_string(NORM_EPS)),
+                 withKey("packed", "false")}));
+
+  layers.push_back(createLayer(
+    "addition",
+    {withKey("name", "layer" + std::to_string(layer_id) + "_post_attention"),
+     withKey("input_layers", input_name + ",layer" + std::to_string(layer_id) +
+                               "_post_attention_norm")}));
+
+  layers.push_back(createLayer(
+    "rms_norm",
+    {withKey("name", "layer" + std::to_string(layer_id) + "pre_ffn_norm"),
+     withKey("input_layers",
+             "layer" + std::to_string(layer_id) + "_post_attention"),
+     withKey("epsilon", std::to_string(NORM_EPS)),
+     withKey("packed", "false")}));
+
+  auto ffn_layer =
+    createMlp(layer_id, DIM, INTERMEDIATE_SIZE,
+              "layer" + std::to_string(layer_id) + "pre_ffn_norm");
+  layers.insert(layers.end(), ffn_layer.begin(), ffn_layer.end());
+
+  layers.push_back(createLayer(
+    "rms_norm",
+    {withKey("name", "layer" + std::to_string(layer_id) + "post_ffn_norm"),
+     withKey("epsilon", std::to_string(NORM_EPS)),
+     withKey("packed", "false")}));
+
+  layers.push_back(createLayer(
+    "addition",
+    {withKey("name", "layer" + std::to_string(layer_id) + "_decoder_output"),
+     withKey("input_layers", "layer" + std::to_string(layer_id) +
+                               "_post_attention,layer" +
+                               std::to_string(layer_id) + "post_ffn_norm")}));
+
+  return layers;
+}
+
+std::vector<LayerHandle>
+Gemma3CausalLM::createAttention(const int layer_id, int seq_len, int n_heads,
+                                int head_dim, std::string query_name,
+                                std::string key_name, std::string value_name) {
+  std::vector<LayerHandle> layers;
+
+  auto Q = "layer" + std::to_string(layer_id) + "_wq";
+  auto Q_norm = "layer" + std::to_string(layer_id) + "_q_norm";
+  auto K = "layer" + std::to_string(layer_id) + "_wk";
+  auto K_norm = "layer" + std::to_string(layer_id) + "_k_norm";
+  auto V = "layer" + std::to_string(layer_id) + "_wv";
+  auto A = "layer" + std::to_string(layer_id) + "_attention";
+  auto O = "layer" + std::to_string(layer_id) + "_attention_out";
+
+  // Q layer
+  std::vector<std::string> q_params = {
```

Comment:
It could be merged QKV layer for the performance reason.

Reply:
I agreed. I'll make optimization PRs for CausalLMs

Reply:
I've applied QKV layer to Gemma3 and Qwen3 on PR #3679 !